lambdas/preview/index.py

```diff
@@ -163,13 +168,61 @@ def extract_vcf(file_):
                 break
 
     # pylint: disable=bad-continuation
-    html = ("<div>"
+    html = ('<div>'
                 "<div data-type='meta'>{}</div>"
                 "<div data-type='header'>{}</div>"
                 "<div data-type='data'>{}</div>"
-            "</div>").format(
-                ''.join(meta) + '<br />',
-                ''.join(header) + '<br />',
+            '</div>').format(
+                ''.join(meta),
+                ''.join(header),
                 ''.join(data))
 
     return html, {}
+
+def extract_txt(file_):
+    """
+    Display first and last lines of file.
+    Uses UNIX utils for speed and convenience (no easy way to tail in Python).
+
+    Args:
+        file_ - file-like
+
+    Throws:
+        CalledProcessError
+    """
+    available = 0
+    ellipsis = False
+    # find out how many lines <= MAX_LINES are available
+    # we need to know this so that head and tail don't overlap
+    # loop finishes in constant time
+    with open(file_.name, 'rb') as txt:
+        for index, line in enumerate(txt, start=1):
+            available = index
+            if available > MAX_LINES:
+                ellipsis = True
+                break
+    # if maxlines == 0, everything still works
+    available = min(available, MAX_LINES)
+    # pylint should never flag for 3.6 but it does
+    headlines = int(available/2) # pylint: disable=old-division
+    taillines = available - headlines
+
+    head = subprocess.check_output(['head', file_.name, '-n', str(headlines)])
+    tail = subprocess.check_output(['tail', file_.name, '-n', str(taillines)])
+    html_buffer = [[], []]
+    for i, lines in enumerate([head, tail]):
+        for line in lines.splitlines():
+            line = line.strip().decode('unicode_escape')
+            html_buffer[i].append(f'<p>{line}</p>')
+
+    # pylint: disable=bad-continuation
+    html = ('<div>'
+                "<div data-type='head'>{}</div>"
```

Comment:
do you want the indents inside the strings?

Reply:
so developer can see HTML indentation but this whole HTML thing is busted and i'm going to refactor

Reply:
Yeah, I meant do you want the indents in the html too so that it looks ok in the dev tools, like:
```
"    <div.."
```
vs
```
    "<div.."
```

If you don't care about the readability/indentation of the out put htlm you could wrap the whole thing in triple quotes instead.

Reply:
i shouldn't have returned HTML in the first place :P (unnecessarily emulating what we did with nbformat). we return JSON now so this is a non-issue, but the problem with newlines in the HTML text is that they get encoded as `\n` and show up in the DOM. an AST or fully server-side-rendered is preferred.